creator-node/src/services/stateMachineManager/CNodeHealthManager.ts

```diff
@@ -0,0 +1,318 @@
+    verboseHealthCheckResp
+  if (
+    thirtyDayRollingSyncSuccessCount &&
+    thirtyDayRollingSyncFailCount &&
+    thirtyDayRollingSyncSuccessCount + thirtyDayRollingSyncFailCount >
+      MINIMUM_ROLLING_SYNC_COUNT &&
+    thirtyDayRollingSyncSuccessCount /
+      (thirtyDayRollingSyncFailCount + thirtyDayRollingSyncSuccessCount) <
+      MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE
+  ) {
+    throw new Error(
+      `Rolling sync data shows that this node fails at a high rate of syncs. Successful syncs=${thirtyDayRollingSyncSuccessCount} || Failed syncs=${thirtyDayRollingSyncFailCount}. Minimum successful sync percentage=${
+        MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE * 100
+      }%`
+    )
+  }
+}
+
+/**
+ * @param node the endpoint to check for health
+ * @param isPrimary whether or not the node is a primary (primaries have a longer grace period)
+ * @return true if the node is either:
+ * - healthy; or
+ * - unhealthy but in a grace period (extra time where it can be unhealthy before being reconfiged)
+ */
+async function isNodeHealthyOrInGracePeriod(node: string, isPrimary: boolean) {
+  const isHealthy = await CNodeHealthManager.isNodeHealthy(node, isPrimary)
+  if (isHealthy) {
+    // If a node ever becomes healthy again and was once marked as unhealthy, remove tracker
+    await CNodeHealthManager.resetEarliestUnhealthyTimestamp(node, isPrimary)
+    return true
+  } else {
+    const failedTimestamp =
+      await CNodeHealthManager.getEarliestFailedHealthCheckTimestamp(
+        node,
+        isPrimary
+      )
+
+    if (failedTimestamp) {
+      // Grace period ends X seconds after the first failed health check
+      const gracePeriodEndDate = new Date(failedTimestamp)
+      gracePeriodEndDate.setSeconds(
+        failedTimestamp.getSeconds() +
+          (isPrimary
+            ? MAX_NUMBER_SECONDS_PRIMARY_REMAINS_UNHEALTHY
+            : MAX_NUMBER_SECONDS_SECONDARY_REMAINS_UNHEALTHY)
+      )
+
+      const now = new Date()
+      if (now >= gracePeriodEndDate) return false
+    } else {
+      await CNodeHealthManager.setHealthCheckTimestampToNow(node, isPrimary)
+    }
+    return true
+  }
+}
+
+async function getEarliestFailedHealthCheckTimestamp(
+  node: string,
+  isPrimary: boolean
+): Promise<Date | null> {
+  const timestampString = await redis.get(
+    `${
+      isPrimary ? REDIS_KEY_PREFIX_PRIMARY : REDIS_KEY_PREFIX_SECONDARY
+    }${node}`
+  )
+  return timestampString ? new Date(timestampString) : null
+}
+
+async function setHealthCheckTimestampToNow(node: string, isPrimary: boolean) {
+  return redis.set(
+    `${
+      isPrimary ? REDIS_KEY_PREFIX_PRIMARY : REDIS_KEY_PREFIX_SECONDARY
+    }${node}`,
+    new Date()
+  )
+}
+
+async function resetEarliestUnhealthyTimestamp(
+  node: string,
+  isPrimary: boolean
+) {
+  return redis.del(
+    `${
+      isPrimary ? REDIS_KEY_PREFIX_PRIMARY : REDIS_KEY_PREFIX_SECONDARY
+    }${node}`
+  )
+}
```

Comment:
these are the only real changes in this file - other changes are copied from the JS version